### ISSUE TITLE
JaaS bundle with local charm & resources error

### ISSUE THREAD

**Issue by johnsca** [canonical/jujucharms.com#526]
Getting the following error when trying to deploy to JaaS but not to a self-hosted controller (with Juju stable, 2.3.5):

```
ERROR cannot deploy bundle: bad file info: negative size
```

Trying to deploy the [canonical-kubernetes bundle](https://jujucharms.com/canonical-kubernetes) with the following overlay to bring in locally built copies of two of the charms for testing:

```yaml
applications:
  kubernetes-master:
    charm: /home/johnsca/juju/envs/dev/builds/kubernetes-master
    resources:
      kubectl:
        kube-apiserver: "/home/johnsca/juju/envs/dev/resources/kube-apiserver.snap"
        kube-controller-manager: "/home/johnsca/juju/envs/dev/resources/kube-controller-manager.snap"
        kube-scheduler: "/home/johnsca/juju/envs/dev/resources/kube-scheduler.snap"
        cdk-addons: "/home/johnsca/juju/envs/dev/resources/cdk-addons.snap"
  kubernetes-worker:
    charm: /home/johnsca/juju/envs/dev/builds/kubernetes-worker
    resources:
      cni-amd64: "/home/johnsca/juju/envs/dev/resources/cni-amd64.tgz"
      cni-s390x: "/home/johnsca/juju/envs/dev/resources/cni-s390x.tgz"
      kubectl: "/home/johnsca/juju/envs/dev/resources/kubectl.snap"
      kubelet: "/home/johnsca/juju/envs/dev/resources/kubelet.snap"
      kube-proxy: "/home/johnsca/juju/envs/dev/resources/kube-proxy.snap"
```

The local resources cover all defined by the charm and are mostly 0-byte placeholders to tell the charms to install from the charm store:

```
$ ls -l resources/
total 15880
-rw-rw-r-- 1 johnsca johnsca        0 Apr  9 11:31 cdk-addons.snap
-rw-rw-r-- 1 johnsca johnsca 16253198 Apr  4 13:09 cni-amd64.tgz
-rw-rw-r-- 1 johnsca johnsca        0 Apr  9 12:01 cni-s390x.tgz
-rw-rw-r-- 1 johnsca johnsca        0 Apr  9 11:31 kube-apiserver.snap
-rw-rw-r-- 1 johnsca johnsca        0 Apr  9 11:31 kube-controller-manager.snap
-rw-rw-r-- 1 johnsca johnsca        0 Apr  9 11:25 kubectl.snap
-rw-rw-r-- 1 johnsca johnsca        0 Apr  9 11:25 kubelet.snap
-rw-rw-r-- 1 johnsca johnsca        0 Apr  9 11:25 kube-proxy.snap
-rw-rw-r-- 1 johnsca johnsca        0 Apr  9 11:31 kube-scheduler.snap
```

The deploy fails on JaaS:

```
$ juju deploy cs:canonical-kubernetes --overlay ./local-k8s.yaml
Located bundle "cs:bundle/canonical-kubernetes-179"
Resolving charm: cs:~containers/easyrsa-33
Resolving charm: cs:~containers/etcd-74
Resolving charm: cs:~containers/flannel-50
Resolving charm: cs:~containers/kubeapi-load-balancer-55
Executing changes:
- upload charm cs:~containers/easyrsa-33 for series xenial
- deploy application easyrsa on xenial using cs:~containers/easyrsa-33
  added resource easyrsa
- set annotations for easyrsa
- upload charm cs:~containers/etcd-74 for series xenial
- deploy application etcd on xenial using cs:~containers/etcd-74
  added resource etcd
  added resource snapshot
- set annotations for etcd
- upload charm cs:~containers/flannel-50 for series xenial
- deploy application flannel on xenial using cs:~containers/flannel-50
  added resource flannel-amd64
  added resource flannel-s390x
- set annotations for flannel
- upload charm cs:~containers/kubeapi-load-balancer-55 for series xenial
- deploy application kubeapi-load-balancer on xenial using cs:~containers/kubeapi-load-balancer-55
- expose kubeapi-load-balancer
- set annotations for kubeapi-load-balancer
- upload charm /home/johnsca/juju/envs/dev/builds/kubernetes-master for series xenial
- deploy application kubernetes-master on xenial using /home/johnsca/juju/envs/dev/builds/kubernetes-master
  added resource cdk-addons
  added resource kube-apiserver
  added resource kube-controller-manager
  added resource kube-scheduler
  added resource kubectl
- set annotations for kubernetes-master
- upload charm /home/johnsca/juju/envs/dev/builds/kubernetes-worker for series xenial
- deploy application kubernetes-worker on xenial using /home/johnsca/juju/envs/dev/builds/kubernetes-worker
ERROR cannot deploy bundle: bad file info: negative size
```

But the exact same overlay, files, and command work with a self-hosted controller:

```
$ juju deploy cs:canonical-kubernetes --overlay ./local-k8s.yaml
Located bundle "cs:bundle/canonical-kubernetes-179"
Resolving charm: cs:~containers/easyrsa-33
Resolving charm: cs:~containers/etcd-74
Resolving charm: cs:~containers/flannel-50
Resolving charm: cs:~containers/kubeapi-load-balancer-55
Executing changes:
- upload charm cs:~containers/easyrsa-33 for series xenial
- deploy application easyrsa on xenial using cs:~containers/easyrsa-33
  added resource easyrsa
- set annotations for easyrsa
- upload charm cs:~containers/etcd-74 for series xenial
- deploy application etcd on xenial using cs:~containers/etcd-74
  added resource etcd
  added resource snapshot
- set annotations for etcd
- upload charm cs:~containers/flannel-50 for series xenial
- deploy application flannel on xenial using cs:~containers/flannel-50
  added resource flannel-amd64
  added resource flannel-s390x
- set annotations for flannel
- upload charm cs:~containers/kubeapi-load-balancer-55 for series xenial
- deploy application kubeapi-load-balancer on xenial using cs:~containers/kubeapi-load-balancer-55
- expose kubeapi-load-balancer
- set annotations for kubeapi-load-balancer
- upload charm /home/johnsca/juju/envs/dev/builds/kubernetes-master for series xenial
- deploy application kubernetes-master on xenial using /home/johnsca/juju/envs/dev/builds/kubernetes-master
  added resource cdk-addons
  added resource kube-apiserver
  added resource kube-controller-manager
  added resource kube-scheduler
  added resource kubectl
- set annotations for kubernetes-master
- upload charm /home/johnsca/juju/envs/dev/builds/kubernetes-worker for series xenial
- deploy application kubernetes-worker on xenial using /home/johnsca/juju/envs/dev/builds/kubernetes-worker
  added resource cni-amd64
  added resource cni-s390x
  added resource kube-proxy
  added resource kubectl
  added resource kubelet
- expose kubernetes-worker
- set annotations for kubernetes-worker
- add relation kubernetes-master:kube-api-endpoint - kubeapi-load-balancer:apiserver
- add relation kubernetes-master:loadbalancer - kubeapi-load-balancer:loadbalancer
- add relation kubernetes-master:kube-control - kubernetes-worker:kube-control
- add relation kubernetes-master:certificates - easyrsa:client
- add relation etcd:certificates - easyrsa:client
- add relation kubernetes-master:etcd - etcd:db
- add relation kubernetes-worker:certificates - easyrsa:client
- add relation kubernetes-worker:kube-api-endpoint - kubeapi-load-balancer:website
- add relation kubeapi-load-balancer:certificates - easyrsa:client
- add relation flannel:etcd - etcd:db
- add relation flannel:cni - kubernetes-master:cni
- add relation flannel:cni - kubernetes-worker:cni
- add unit easyrsa/0 to new machine 0
- add unit etcd/0 to new machine 1
- add unit etcd/1 to new machine 2
- add unit etcd/2 to new machine 3
- add unit kubeapi-load-balancer/0 to new machine 4
- add unit kubernetes-master/0 to new machine 5
- add unit kubernetes-worker/0 to new machine 6
- add unit kubernetes-worker/1 to new machine 7
- add unit kubernetes-worker/2 to new machine 8
Deploy of bundle completed.
```

Reported here per @mitechie

**Comment by rogpeppe** [comment]
I can't reproduce this locally. Here's my attempt at reproducing the issue. I've uploaded the issue526.tgz file to http://rogpeppe-scratch.s3.amazonaws.com/issue526.tgz . It has SHA256 sum de81934de02dc1bb2245407871b54c3fbfbc24c540a2adfcd4639b340b661ad1.

	% cd /tmp
	% tar xzf issue526.tgz
	% cd issue526
	% juju add-model jaastest2 aws
	Added 'jaastest2' model on aws/eu-west-1 with credential 'ec2' for user 'rogpeppe'
	% juju deploy cs:canonical-kubernetes --overlay ./local-k8s.yaml
	Located bundle "cs:bundle/canonical-kubernetes-218"
	Resolving charm: cs:~containers/easyrsa-50
	Resolving charm: cs:~containers/etcd-96
	Resolving charm: cs:~containers/flannel-66
	Resolving charm: cs:~containers/kubeapi-load-balancer-69
	Executing changes:
	- upload charm cs:~containers/easyrsa-50 for series xenial
	- deploy application easyrsa on xenial using cs:~containers/easyrsa-50
	  added resource easyrsa
	- set annotations for easyrsa
	- upload charm cs:~containers/etcd-96 for series xenial
	- deploy application etcd on xenial using cs:~containers/etcd-96
	  added resource etcd
	  added resource snapshot
	- set annotations for etcd
	- upload charm cs:~containers/flannel-66 for series xenial
	- deploy application flannel on xenial using cs:~containers/flannel-66
	  added resource flannel-amd64
	  added resource flannel-arm64
	  added resource flannel-s390x
	- set annotations for flannel
	- upload charm cs:~containers/kubeapi-load-balancer-69 for series xenial
	- deploy application kubeapi-load-balancer on xenial using cs:~containers/kubeapi-load-balancer-69
	- expose kubeapi-load-balancer
	- set annotations for kubeapi-load-balancer
	- upload charm /tmp/issue526/builds/kubernetes-master for series xenial
	- deploy application kubernetes-master on xenial using /tmp/issue526/builds/kubernetes-master
	  added resource cdk-addons
	  added resource kube-apiserver
	  added resource kube-controller-manager
	  added resource kube-scheduler
	  added resource kubectl
	- set annotations for kubernetes-master
	- upload charm /tmp/issue526/builds/kubernetes-worker for series xenial
	- deploy application kubernetes-worker on xenial using /tmp/issue526/builds/kubernetes-worker
	  added resource cni-amd64
	  added resource cni-arm64
	  added resource cni-s390x
	  added resource kube-proxy
	  added resource kubectl
	  added resource kubelet
	- expose kubernetes-worker
	- set annotations for kubernetes-worker
	- add relation kubernetes-master:kube-api-endpoint - kubeapi-load-balancer:apiserver
	- add relation kubernetes-master:loadbalancer - kubeapi-load-balancer:loadbalancer
	- add relation kubernetes-master:kube-control - kubernetes-worker:kube-control
	- add relation kubernetes-master:certificates - easyrsa:client
	- add relation etcd:certificates - easyrsa:client
	- add relation kubernetes-master:etcd - etcd:db
	- add relation kubernetes-worker:certificates - easyrsa:client
	- add relation kubernetes-worker:kube-api-endpoint - kubeapi-load-balancer:website
	- add relation kubeapi-load-balancer:certificates - easyrsa:client
	- add relation flannel:etcd - etcd:db
	- add relation flannel:cni - kubernetes-master:cni
	- add relation flannel:cni - kubernetes-worker:cni
	- add unit easyrsa/0 to new machine 0
	- add unit etcd/0 to new machine 1
	- add unit etcd/1 to new machine 2
	- add unit etcd/2 to new machine 3
	- add unit kubeapi-load-balancer/0 to new machine 4
	- add unit kubernetes-master/0 to new machine 5
	- add unit kubernetes-master/1 to new machine 6
	- add unit kubernetes-worker/0 to new machine 7
	- add unit kubernetes-worker/1 to new machine 8
	- add unit kubernetes-worker/2 to new machine 9
	Deploy of bundle completed.
	%

**Comment by johnsca** [comment]
I can no longer reproduce this issue, either.  It may have been resolved by subsequent updates to Juju, JAAS, or the charms.

I should note that the only other time I've seen the "negative file size" error is when something strange was going on with a charm revision in the store (e.g., a revision being pulled out of a channel) which led to a resource being referenced that didn't technically exist, and I haven't seen that in some time.  At a guess, maybe there was a fix for some corner case around that which made it into my bootstrapped controller but wasn't available yet in the JAAS controller at the time but which has now been released there.

I also noticed that the overlay YAML is actually malformed (the resources for `kubernetes-master` is missing `kubectl` and has all the other resources nested under that).  This seems like it ought to be rejected by the controller but somehow works anyway, so I don't know whether it's related to what I had been seeing or not.

At any rate, closing this issue because it's some combination of user error and already resolved.
